### PR TITLE
[FSDP] Fix train -> EMA -> eval with mixed precision

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #106858
* #106857

This fixes a pretty vicious bug relating to `SHARD_GRAD_OP`, mixed precision, EMA, and eval.

**Bug Explanation**
The model has a main module and an EMA module, where the main module is used for training and the EMA module is used for eval. The model has FSDP's fp16 mixed precision enabled. The flow consists of (1) training forward/backward/optimizer -> (2) EMA update (copy main module to EMA module) -> eval forward in `torch.no_grad()`, where this repeats for many iterations.

Consider the _second_ iteration.
- From the first iteration's eval forward, the EMA module has the fp16 unsharded parameters in memory (not freed due to `SHARD_GRAD_OP`).
- In this second iteration's step (2), we perform the EMA update under the `summon_full_params()` context, where FSDP specially forces full precision.  This means that the EMA module now uses fp32 unsharded parameters, distinct from the fp16 unsharded parameters still in memory. The EMA update modifies those fp32 parameters, and upon exiting the context, FSDP correctly writes the modifications back to the fp32 sharded parameters.
- In the second iteration's step (3) (eval forward), FSDP checks whether it needs to run the unshard op (including all-gather) but sees it does not since the fp16 unsharded parameters are still in memory. Thus, FSDP uses those fp16 unsharded parameters directly without all-gather. However, these fp16 unsharded parameters are stale and do not include the EMA update!
- In other words, at this point, the fp32 sharded parameters are correct, the fp16 unsharded parameters are stale, and FSDP chooses _not_ to re-all-gather since the fp16 unsharded parameters are in memory.

**Fix Explanation**
This PR fixes this by freeing the fp16 unsharded parameters if they are still allocated when forcing full precision, i.e. using fp32 unsharded parameters in `summon_full_params()`. This ensures that any modifications written back to the fp32 sharded parameters will be persisted via the next all-gather.

### CODE REVIEW

File: torch/distributed/fsdp/flat_param.py [hunk 1]
```diff
@@ -1293,6 +1293,14 @@ def _get_padded_unsharded_flat_param(self) -> torch.Tensor:
                 unsharded_flat_param.dtype != self._fwd_bwd_param_dtype,
                 f"Expects full precision but got {self._fwd_bwd_param_dtype}",
             )
+            # For no-reshard-after-forward strategies, `_full_param_padded` may
+            # still be allocated from a previous forward. As we are forcing
+            # full precision here, the full-precision unsharded copy may be
+            # modified, invalidating the existing low-precision unsharded copy,
+            # so we should free it here to ensure a new all-gather for the next
+            # forward/backward computation to persist the modifications.
+            if flat_param._full_param_padded.untyped_storage().size() > 0:
```
Comment (kumpera) [inline torch/distributed/fsdp/flat_param.py:1302]: Won't this regress the scenario where people use `summon_full_params(writeback=False)`?

**Review by kumpera — APPROVED** [review]
LGTM.
My only concern is whether this fix might regress perf when using `writeback=False`.

**Comment by awgu** [comment]
> LGTM. My only concern is whether this fix might regress perf when using `writeback=False`.

That is a fair point! I think we would prefer to free in this case anyway to avoid increasing memory usage so much (having both the fp32 and fp16 unsharded parameters in memory). I am not too aware of who uses `writeback=False` in practice, so we can revisit this if we have such a case that cares about not re-all-gathering.

**Comment by awgu** [comment]
Failure is unrelated (unstable):
[linux-focal-rocm5.6-py3.8 / test (default, 1, 3, linux.rocm.gpu, unstable)](https://hud.pytorch.org/pr/pytorch/pytorch/106858#15757954657)

**Comment by ezyang** [comment]
Something I don't quite understand: because we never call backward() on the EMA module, we never reshard the EMA weights (as requested by the user.) But isn't this... weird? Like, did you really want to keep the EMA weights around while you train the main module?

The strategy taken by this PR doesn't seem complete to me. You rely on explicit invalidation when `_get_padded_unsharded_flat_param` is requested. But is this the *only* way to mutate the "canonical" (full precision) copy of the weights? If it isn't, then you still can potentially have the same bug, where you have a stale cache floating around. If you can deal with the cost, a more comprehensive fix might be to associate the canonical copy of the parameters with a version counter (tracked automatically for you by PyTorch), and whenever you generate a cached copy of the parameters, you save the version counter it corresponds to. If upon further access, you notice that they mismatch, you trigger a refetch. Then you have an optimization problem which is how to make sure this refetch is prefetched appropriately, but this can mostly be done by sprinkling in "is the cache invalid yet?" tests whenever you would have normally prefetched.

**Comment by awgu** [comment]
> But is this the only way to mutate the "canonical" (full precision) copy of the weights?

I agree that this is brittle, but at least for now, this is the only way. (I am not sure how there could be another way to get the full precision unsharded copy even in the future.) What you said makes sense though, and I will think about this more.

> because we never call backward() on the EMA module, we never reshard the EMA weights (as requested by the user.) But isn't this... weird? Like, did you really want to keep the EMA weights around while you train the main module?

In theory, the user knows a priori that the EMA module will only be used in forward passes. In that case, if the user wants to free the EMA weights, then the user should use `FULL_SHARD` strategy that does free the weights after forward. I would take it that using `SHARD_GRAD_OP` would mean the user is intentionally not freeing the weights after forward.

**Comment by ezyang** [comment]
> In theory, the user knows a priori that the EMA module will only be used in forward passes. In that case, if the user wants to free the EMA weights, then the user should use FULL_SHARD strategy that does free the weights after forward. I would take it that using SHARD_GRAD_OP would mean the user is intentionally not freeing the weights after forward.

I doubt the user had any such complicated thought process. They probably were thinking about what strategy they wanted for the regular forward-backward training and forgot that the policy also applies to the EMA weights... oops! In other words, I expected that they would apply SHARD_GRAD_OP for the trainable module, and FULL_SHARD for the eval-only module. But if this is actually the normal thing to want, it would be nice for SHARD_GRAD_OP to decay to FULL_SHARD when we know a backward cannot happen. This could be detected via requires_grad, perhaps.
